### PR TITLE
package-indexer: rpm repo

### PR DESCRIPTION
**Pull request by alltilla**
Example run: https://github.com/alltilla/axosyslog/actions/runs/12032052926/job/33543787289

You can try it out based on the README.md and this repo file:
```
$ cat /etc/yum.repos.d/axosyslog.repo 
[axosyslog]
name=AxoSyslog
baseurl=https://axosyslog.alltilla.com/rpm/nightly/fedora-$releasever/$basearch
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=https://axosyslog.alltilla.com/alltilla-code-signing-pub.asc
```

### CODE REVIEW

File: packaging/package-indexer/indexer/rpm_indexer.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            f"%__gpg {shutil.which('gpg')}\n"
+            f"%__gpg_sign_cmd {sign_cmd}\n"
+        )
+
+        for rpm_file in indexed_dir.rglob("*.rpm"):
+            self._log_info("Signing rpm.", rpm_file=str(rpm_file))
+            utils.execute_command(
+                ["rpm", "--macros", str(rpmmacros_path), "--addsign", str(rpm_file.absolute())],
+                input=self.__gpg_key_passphrase,
+            )
```
Comment (alltilla) [inline packaging/package-indexer/indexer/rpm_indexer.py:114]: gpg key with passphrase does not work.

probably we need to use a gpg-agent session, so passphrase input is not needed for each sign.

**Comment by alltilla** [comment]
Example `.repo` file:
```
[axosyslog]
name=AxoSyslog
baseurl=http://localhost:8080/rpm/stable/fedora-$releasever/$basearch
enabled=1
gpgcheck=1
repo_gpgcheck=1
gpgkey=http://localhost:8080/key.pub
```

File: packaging/package-indexer/indexer/rpm_indexer.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+            cdn=cdn,
+        )
+
+    def __move_files_from_incoming_to_indexed(self, incoming_dir: Path, indexed_dir: Path) -> None:
+        for file in filter(lambda path: path.is_file(), incoming_dir.rglob("*")):
+            relative_path = file.relative_to(incoming_dir)
+            platform = relative_path.parent
+            file_name = relative_path.name
+
+            new_path = Path(indexed_dir, platform, "x86_64", file_name)
```
Comment (MrAnno) [inline packaging/package-indexer/indexer/rpm_indexer.py:67]: Can we extract the arch string to a global constant? (We are not far away from supporting aarch64)

File: packaging/package-indexer/indexer/rpm_indexer.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+        env = {"GNUPGHOME": str(gnupghome_dir), "GPG_TTY": ""}
+
+        for rpm_file in indexed_dir.rglob("*.rpm"):
+            command = ["rpm", "--macros", str(rpmmacros_path), "--addsign", str(rpm_file.absolute())]
+
+            self._log_info("Signing rpm.", rpm_file=str(rpm_file))
+            utils.execute_command(command, env=env)
+
+    def __create_repomd_xml_asc_file(self, gnupghome_dir: Path, repo_path: Path) -> None:
```
Comment (MrAnno) [inline packaging/package-indexer/indexer/rpm_indexer.py:114]: The naming around RPM repos confuses me.
Is this yum-specific? Does the same format apply to `dnf` and `zypper`?

File: packaging/package-indexer/indexer/rpm_indexer.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            cdn=cdn,
+            gpg_key_path=gpg_key_path,
+            gpg_key_passphrase=gpg_key_passphrase,
+        )
+
+    def __get_pkg_timestamps_in_dir(self, dir: Path) -> List[str]:
+        timestamp_regexp = re.compile(r"\+([^_]+)_")
+        pkg_timestamps: List[str] = []
+
+        for deb_file in dir.rglob("syslog-ng-devel*.rpm"):
```
Comment (MrAnno) [inline packaging/package-indexer/indexer/rpm_indexer.py:277]: axosyslog?

File: packaging/package-indexer/indexer/rpm_indexer.py (hunk 3 above)
Comment (alltilla) [reply packaging/package-indexer/indexer/rpm_indexer.py:114]: As far as I can understand the only structure that is needed is to have a dir with `.rpm`s, and a `repodata` subdir, for the meta files (`createrepo` makes this for us).

On how you find the dir for your OS and arch:
The dir structure does not matter (e.g. `rpm/nightly/almalinux-8/x86_64/`), but it is useful to name them in a way, that the yum variables resolve to: https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/6/html/deployment_guide/sec-using_yum_variables#sec-Using_Yum_Variables

I tested with `dnf`, and it uses the same `/etc/yum.repos.d` dir, and works. Did not test `zypper`, though.

So I guess parts of this is yum specific, but not sure how much.

```
# dnf install axosyslog                                                                                                                                                                         
AlmaLinux 9 - CRB                                                                                                                                                                    3.7 MB/s | 2.7 MB     00:00     
Extra Packages for Enterprise Linux 9 - x86_64                                                                                                                                        15 MB/s |  23 MB     00:01     
Dependencies resolved.                                                                                                                                                                                               
=====================================================================================================================================================================================================================
 Package                                             Architecture                           Version                                                                  Repository                                 Size 
=====================================================================================================================================================================================================================
Installing:
 axosyslog                                           x86_64                                 0.8.1.18143.g7f5f3f5-2+20241126T135514                                   axosyslog                                 1.1 M
Installing dependencies:
 ivykis                                              x86_64                                 0.42.4-7.el9                                                             epel                                       46 k
 libnet                                              x86_64                                 1.2-7.el9                                                                appstream                                  57 k
 logrotate                                           x86_64                                 3.18.0-8.el9                                                             baseos                                     74 k
Installing weak dependencies:
 axosyslog-logrotate                                 x86_64                                 0.8.1.18143.g7f5f3f5-2+20241126T135514                                   axosyslog                                 8.6 k

Transaction Summary
=====================================================================================================================================================================================================================
Install  5 Packages

Total download size: 1.3 M
...
```

File: packaging/package-indexer/indexer/rpm_indexer.py (hunk 4 above)
Comment (alltilla) [reply packaging/package-indexer/indexer/rpm_indexer.py]: Thanks, nice catch.

An example file name (0.8.1 comes from not having tags in my fork):
```
axosyslog-devel-0.8.1.18105.g4ca9958-2+20241117T182608.x86_64.rpm
```

The pattern is:
```
\+([^.]+).
```

Will match like:
image

File: packaging/package-indexer/indexer/rpm_indexer.py (hunk 2 above)
Comment (alltilla) [reply packaging/package-indexer/indexer/rpm_indexer.py]: Yes, done.

File: packaging/package-indexer/indexer/rpm_indexer.py (hunk 3 above)
Comment (MrAnno) [reply packaging/package-indexer/indexer/rpm_indexer.py:115]: Thanks.

The man of createrepo says `repomd (xml-based rpm metadata)`, which also clears a few things for me.

**Review by MrAnno — APPROVED** [review]

**Pull request merged**